<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>PRJNA517180 • enriquepresa</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="PRJNA517180">
<meta property="og:description" content="enriquepresa">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">enriquepresa</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/PRJNA517180.html">PRJNA517180</a>
    </li>
    <li>
      <a href="../articles/tarea1.html">tarea1</a>
    </li>
    <li>
      <a href="../articles/tarea2.html">tarea2</a>
    </li>
    <li>
      <a href="../articles/tarea3_2.html">tarea3_2</a>
    </li>
    <li>
      <a href="../articles/tarea4_1.html">tarea4_1</a>
    </li>
    <li>
      <a href="../articles/tarea4_2.html">tarea4_2</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>PRJNA517180</h1>
                        <h4 class="author">Enrique Presa</h4>
            
            <h4 class="date">23 junio, 2020</h4>
      
      
      <div class="hidden name"><code>PRJNA517180.rmd</code></div>

    </div>

    
    
<div id="descarga-de-datos-de-prjna517180" class="section level1">
<h1 class="hasAnchor">
<a href="#descarga-de-datos-de-prjna517180" class="anchor"></a>Descarga de datos de PRJNA517180</h1>
<p>Los información de los datos de RNA-seq que se van a descargar se pueden encontrar en <a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA517180" class="uri">https://www.ncbi.nlm.nih.gov/bioproject/PRJNA517180</a> En el estudio PRJNA517180 se hace un RNA-seq en individuos de <em>C. elegans</em> para estudiar la importancia del gen ztf-11 en la neurogénesis: la pérdida de ztf-11 da lugar a la sobreexpresión de genes no neuronales y se reduce la neurogénesis. En este experimento se utilizan dos grupos: un grupo control con un plásmido L4440 que no genera cambios fisiológicos, y por otra parte un grupo en el que el gen ztf-11 es silenciado por RNA de interferencia.</p>
<div id="descarga-de-todos-los-ficheros-en-formato-sra" class="section level2">
<h2 class="hasAnchor">
<a href="#descarga-de-todos-los-ficheros-en-formato-sra" class="anchor"></a>Descarga de todos los ficheros en formato SRA</h2>
<p>Se ha descargado manualmente un fichero de texto denominado <em>SraAccList.txt</em> con los <em>accession IDs SRR</em> para descargar todos los datos en formato SRA. Para que sea reproducible desde el primer momento, se va a construir este fichero. Con la opción <em>prefetch</em> del <em>SRA Toolkit</em> se descargan todos los ficheros con formato sra de la lista del fichero anterior. También se guarda el directorio de trabajo actual para utilizarlo más adelante.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="no">wd</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span>()</pre></body></html></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb2-1" title="1"><span class="bu">echo</span> -e <span class="st">"SRR8489772\nSRR8489773\nSRR8489774\nSRR8489775\nSRR8489776\nSRR8489777\nSRR8489778\nSRR8489779"</span> <span class="op">&gt;</span> SraAccList.txt</a>
<a class="sourceLine" id="cb2-2" title="2"></a>
<a class="sourceLine" id="cb2-3" title="3"><span class="ex">prefetch</span> --option-file SraAccList.txt</a></code></pre></div>
</div>
<div id="conversión-a-formato-fastq" class="section level2">
<h2 class="hasAnchor">
<a href="#conversi%C3%B3n-a-formato-fastq" class="anchor"></a>Conversión a formato fastq</h2>
<p>Una vez descargados, <em>SRA Toolkit</em> hace un subdirectorio en HOME, así que se cambia el directorio de trabajo ahí y se utiliza <em>fastq-dump</em> con todas las secuencias para transformarlas a ficheros fastq. En este estudio se utilizaron <em>paired-reads</em>, así que hay que utilizar la opción <em>split files</em>. Se borran los ficheros SRA.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" title="1"><span class="bu">cd</span> /<span class="va">$HOME</span>/ncbi/public/sra</a>
<a class="sourceLine" id="cb3-2" title="2"><span class="kw">for</span> <span class="ex">fichero</span> in <span class="kw">`</span><span class="fu">ls</span> .<span class="kw">`;</span> <span class="kw">do</span> <span class="ex">fastq-dump</span> -I --split-files <span class="va">$fichero</span><span class="kw">;</span> <span class="kw">done</span></a>
<a class="sourceLine" id="cb3-3" title="3"><span class="fu">rm</span> *.sra</a></code></pre></div>
</div>
<div id="descarga-del-genoma-de-referencia" class="section level2">
<h2 class="hasAnchor">
<a href="#descarga-del-genoma-de-referencia" class="anchor"></a>Descarga del genoma de referencia</h2>
<p>Por otra parte, hay que descargar el genoma de referencia a utilizar. En este estudio se ha utilizado el genoma de <em>C. elegans</em>. Por comodidad, se descargará en el directorio del que cuelga el directorio con los ficheros fastq.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb4-1" title="1"><span class="bu">cd</span> /<span class="va">$HOME</span>/ncbi/public</a>
<a class="sourceLine" id="cb4-2" title="2"><span class="fu">wget</span> ftp://igenome:G3nom3s4u@ussd-ftp.illumina.com/Caenorhabditis_elegans/Ensembl/WBcel235/Caenorhabditis_elegans_Ensembl_WBcel235.tar.gz</a>
<a class="sourceLine" id="cb4-3" title="3"><span class="fu">gzip</span> -d Caenorhabditis_elegans_Ensembl_WBcel235.tar.gz</a>
<a class="sourceLine" id="cb4-4" title="4"><span class="fu">tar</span> xvf Caenorhabditis_elegans_Ensembl_WBcel235.tar</a></code></pre></div>
</div>
<div id="mapeo-de-reads-con-el-mapeador-bowtie2" class="section level2">
<h2 class="hasAnchor">
<a href="#mapeo-de-reads-con-el-mapeador-bowtie2" class="anchor"></a>Mapeo de reads con el mapeador Bowtie2</h2>
<p>Para mapear, se utiliza Bowtie2. Para ello, del directorio del que cuelga el subdirectorio con el genoma de referencia y el subdirectorio con las reads en formato fastq, se crea otro subdirectorio denominado <em>aligned</em> en el que se incorporarán los ficheros SAM generados. Por otra parte, se eliminan los ficheros fastq.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb5-1" title="1"><span class="bu">cd</span> /<span class="va">$HOME</span>/ncbi/public</a>
<a class="sourceLine" id="cb5-2" title="2"><span class="fu">mkdir</span> aligned</a>
<a class="sourceLine" id="cb5-3" title="3"></a>
<a class="sourceLine" id="cb5-4" title="4"><span class="kw">for</span> <span class="ex">fichero</span> in <span class="kw">`</span><span class="fu">ls</span> sra/ <span class="kw">|</span> <span class="fu">grep</span> -o -E ^[A-Z0-9]+ <span class="kw">|</span> <span class="fu">sort</span> -u<span class="kw">`;</span> <span class="kw">do</span> <span class="ex">bowtie2</span> -x Caenorhabditis_elegans/Ensembl/WBcel235/Sequence/Bowtie2Index/genome -1 sra/<span class="kw">`</span><span class="bu">echo</span> <span class="va">$fichero</span><span class="kw">`</span>_1.fastq -2 sra/<span class="kw">`</span><span class="bu">echo</span> <span class="va">$fichero</span><span class="kw">`</span>_2.fastq -S aligned/<span class="kw">`</span><span class="bu">echo</span> <span class="va">$fichero</span><span class="kw">`</span>.sam<span class="kw">;done</span></a>
<a class="sourceLine" id="cb5-5" title="5"><span class="fu">rm</span> sra/*.fastq</a></code></pre></div>
</div>
<div id="conversión-a-ficheros-bam" class="section level2">
<h2 class="hasAnchor">
<a href="#conversi%C3%B3n-a-ficheros-bam" class="anchor"></a>Conversión a ficheros bam</h2>
<p>Se pasa a formato BAM gracias a <em>Samtools</em> y después también se borran los ficheros SAM. Más adelante, para hacer los conteos de reads mapeados, se necesita un fichero de texto con el nombre de todos los ficheros BAM generados.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb6-1" title="1"><span class="bu">cd</span> /<span class="va">$HOME</span>/ncbi/public/aligned</a>
<a class="sourceLine" id="cb6-2" title="2"></a>
<a class="sourceLine" id="cb6-3" title="3"><span class="kw">for</span> <span class="ex">fichero</span> in <span class="kw">`</span><span class="fu">ls</span> . <span class="kw">|</span> <span class="fu">grep</span> -o -E <span class="st">"[A-Z]+[0-9]+"</span><span class="kw">`;</span> <span class="kw">do</span> <span class="ex">samtools</span> view -bS <span class="kw">`</span><span class="bu">echo</span> <span class="va">$fichero</span><span class="kw">`</span>.sam <span class="kw">|</span> <span class="ex">samtools</span> sort -o <span class="kw">`</span><span class="bu">echo</span> <span class="va">$fichero</span><span class="kw">`</span>.bam<span class="kw">;</span> <span class="fu">rm</span> <span class="kw">`</span><span class="bu">echo</span> <span class="va">$fichero</span><span class="kw">`</span>.sam<span class="kw">;</span> <span class="kw">done</span></a>
<a class="sourceLine" id="cb6-4" title="4"></a>
<a class="sourceLine" id="cb6-5" title="5"><span class="fu">ls</span> *.bam <span class="op">&gt;</span> bamfiles.txt</a></code></pre></div>
</div>
<div id="conteos-de-reads-alineadas" class="section level2">
<h2 class="hasAnchor">
<a href="#conteos-de-reads-alineadas" class="anchor"></a>Conteos de reads alineadas</h2>
<p>Para ello ya se utiliza R. En concreto los paquetes <em>Rsamtools</em>, <em>GenomicFeatures</em> y <em>GenomicAlignments</em>. En el <em>colData</em> se meten las variables fenotípicas. En este experimento solo será necesario incorporar el SRR bajo el nombre de columna <em>Run</em> y por otra parte el tipo de muestra: si bien el individuo era un mutante knockdown del gen ztf-11 o era individuo del grupo control, denominado según el vector control L4440.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="kw pkg">pacman</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_load.html">p_load</a></span>(<span class="no">Rsamtools</span>, <span class="no">GenomicFeatures</span>, <span class="no">GenomicAlignments</span>)
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span>(<span class="st">"~/ncbi/public"</span>)
<span class="no">gtfFile</span> <span class="kw">=</span> <span class="st">"Caenorhabditis_elegans/Ensembl/WBcel235/Annotation/Genes/genes.gtf"</span>
<span class="no">txdb</span> <span class="kw">=</span> <span class="fu">makeTxDbFromGFF</span>(<span class="no">gtfFile</span>, <span class="kw">format</span><span class="kw">=</span><span class="st">"gtf"</span>)
<span class="no">genes</span> <span class="kw">=</span> <span class="fu">exonsBy</span>(<span class="no">txdb</span>, <span class="kw">by</span><span class="kw">=</span><span class="st">"gene"</span>)

<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span>(<span class="st">"~/ncbi/public/aligned"</span>)
<span class="no">dirActualData</span> <span class="kw">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span>(),<span class="st">"/"</span>,<span class="kw">sep</span><span class="kw">=</span><span class="st">""</span>)
<span class="no">sampleTableSingle</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="st">"bamfiles.txt"</span>)
<span class="no">fls</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="no">dirActualData</span>,<span class="no">sampleTableSingle</span>[,<span class="fl">1</span>],<span class="kw">sep</span><span class="kw">=</span><span class="st">""</span>)
<span class="no">bamLst</span> <span class="kw">=</span> <span class="fu">BamFileList</span>(<span class="no">fls</span>, <span class="kw">index</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">character</a></span>(),<span class="kw">yieldSize</span><span class="kw">=</span><span class="fl">100000</span>,<span class="kw">obeyQname</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="no">PRJNA517180</span> <span class="kw">=</span> <span class="fu">summarizeOverlaps</span>(<span class="kw">features</span> <span class="kw">=</span> <span class="no">genes</span>,<span class="kw">read</span><span class="kw">=</span><span class="no">bamLst</span>,
   <span class="kw">mode</span><span class="kw">=</span><span class="st">"Union"</span>,
      <span class="kw">singleEnd</span><span class="kw">=</span><span class="fl">FALSE</span>,
      <span class="kw">ignore.strand</span><span class="kw">=</span><span class="fl">TRUE</span>,
      <span class="kw">fragments</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="no">Run</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"SRR8489772"</span>, <span class="st">"SRR8489773"</span>, <span class="st">"SRR8489774"</span>, <span class="st">"SRR8489775"</span>, <span class="st">"SRR8489776"</span>, <span class="st">"SRR8489777"</span>, <span class="st">"SRR8489778"</span>, <span class="st">"SRR8489779"</span>)

<span class="no">type</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">0</span>, <span class="fl">4</span>), <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">1</span>, <span class="fl">4</span>))
<span class="no">type</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">type</span>, <span class="kw">levels</span><span class="kw">=</span><span class="fl">0</span>:<span class="fl">1</span>, <span class="kw">labels</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"L4440"</span>, <span class="st">"ztf_11_KD"</span>))
<span class="fu">colData</span>(<span class="no">PRJNA517180</span>) <span class="kw">=</span> <span class="fu">DataFrame</span>(<span class="no">Run</span>, <span class="no">type</span>)</pre></body></html></div>
</div>
</div>
<div id="generación-de-un-rowdata" class="section level1">
<h1 class="hasAnchor">
<a href="#generaci%C3%B3n-de-un-rowdata" class="anchor"></a>Generación de un rowData</h1>
<p>Antes de hacer el informe, es necesario hacer un <em>rowData</em> con diferentes identificadores de todos los genes. Esto será útil más adelante para hacer los informes con diferentes URLs a bases de datos con <em>ReportingTools</em>, y para el proceso de sobre-representación. Después, se guardará el SummarizedExperiment en un fichero Rdata.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="kw pkg">pacman</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_load.html">p_load</a></span>(<span class="no">AnnotationDbi</span>, <span class="no">org.Ce.eg.db</span>)
<span class="no">annot</span> <span class="kw">=</span> <span class="kw pkg">AnnotationDbi</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">select</a></span>(<span class="no">org.Ce.eg.db</span>, <span class="kw">keys</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="fu">assay</span>(<span class="no">PRJNA517180</span>)),
                              <span class="kw">column</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"SYMBOL"</span>, <span class="st">"ENTREZID"</span>, <span class="st">"ENSEMBL"</span>), <span class="kw">keytype</span><span class="kw">=</span><span class="st">"WORMBASE"</span>)
<span class="no">uniq</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="fu">assay</span>(<span class="no">PRJNA517180</span>)),<span class="no">annot</span>[,<span class="fl">1</span>])
<span class="fu">rowData</span>(<span class="no">PRJNA517180</span>)<span class="kw">=</span><span class="no">annot</span>[<span class="no">uniq</span>,]
<span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span>(<span class="no">PRJNA517180</span>, <span class="kw">file</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">wd</span>,<span class="st">"/PRJNA517180.rda"</span>))</pre></body></html></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Enrique Presa.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
